### PR TITLE
[Delete] /api/problems/{problemId}/testcases/{testcaseId} #310

### LINKED ISSUE #310: [Delete] /api/problems/{problemId}/testcases/{testcaseId}
`/api/problems/{problemId}/testcases/{testcaseId}`

### PR DESCRIPTION
**Pull request by ShawnSWu**
Resolved #310

### CODE REVIEW

File: Spring-Boot/Spring-Boot-Problem/src/main/java/tw/waterball/judgegirl/springboot/problem/controllers/ProblemController.java [hunk 1]
```diff
@@ -95,6 +96,16 @@ public ProblemView getProblem(@RequestHeader(value = "Authorization", required =
...
+    @DeleteMapping("/{problemId}/testcases/{testcaseId}")
+    public void deleteTestcase(@PathVariable int problemId,
+                               @PathVariable String testcaseId) {
+        DeleteTestcaseUseCase.Request request = DeleteTestcaseUseCase.Request.builder()
+                .problemId(problemId)
+                .testcaseId(testcaseId)
+                .build();
+        deleteTestcaseUseCase.execute(request);
+    }
+
```
Comment (Johnny850807) [inline Spring-Boot/Spring-Boot-Problem/src/main/java/tw/waterball/judgegirl/springboot/problem/controllers/ProblemController.java:108]: do **AUTH** (admin-only), see other methods for reference

File: Spring-Boot/Spring-Boot-Problem/src/test/java/tw/waterball/judgegirl/springboot/problem/controllers/ProblemControllerTest.java [hunk 2]
```diff
@@ -464,6 +464,24 @@ void GivenProblemSaved_WhenAddNewTestcase_ThenShouldAddSuccessfully() throws Exc
...
     }
 
+    private void deleteTestCase(int problemId, String testCaseId) throws Exception {
+        mockMvc.perform(delete(API_PREFIX + "/{problemId}/testcases/{testcaseId}", problemId, testCaseId)
+                .accept(MediaType.APPLICATION_JSON))
+                .andExpect(status().isOk());
+    }
+
+    @Test
+    void GivenProblemSaved_WhenDeleteTestcase_ThenShouldDeleteSuccessfully() throws Exception {
```
Comment (Johnny850807) [inline Spring-Boot/Spring-Boot-Problem/src/test/java/tw/waterball/judgegirl/springboot/problem/controllers/ProblemControllerTest.java:474]: `GivenProblemSavedWithTestcases_WhenDeleteOneOfItsTestcase_ThenShouldDeleteSuccessfully`

File: Spring-Boot/Spring-Boot-Problem/src/test/java/tw/waterball/judgegirl/springboot/problem/controllers/ProblemControllerTest.java [hunk 3]
```diff
@@ -464,6 +464,24 @@ void GivenProblemSaved_WhenAddNewTestcase_ThenShouldAddSuccessfully() throws Exc
...
+        int problemId = 1;
+        Problem problem = saveProblemAndGet(problemId);
+
+        var expectDeletedTestcase = problem.getTestcases().get(0);
+        deleteTestCase(problem.getId(), expectDeletedTestcase.getId());
+
+        var actualTestcases = getProblem(withToken(adminToken), problemId).getTestcases();
+        assertEquals(problem.getTestcases().size() - 1, actualTestcases.size());
```
Comment (Johnny850807) [inline Spring-Boot/Spring-Boot-Problem/src/test/java/tw/waterball/judgegirl/springboot/problem/controllers/ProblemControllerTest.java:482]: validate more strictly --> make sure the correct testcase is deleted

File: Domain/Problem/src/main/java/tw/waterball/judgegirl/problem/domain/usecases/DeleteTestcaseUseCase.java [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+        this.testCaseRepository = testCaseRepository;
+    }
+
+    public void execute(Request request) throws NotFoundException {
+        testCaseRepository.deleteTestcaseById(request.problemId, request.testcaseId);
+    }
+
+
+    @AllArgsConstructor
+    @Builder
```
Comment (Johnny850807) [inline Domain/Problem/src/main/java/tw/waterball/judgegirl/problem/domain/usecases/DeleteTestcaseUseCase.java:29]: don't use Builder when there are just 1~3 parameters

File: Domain/Problem/src/main/java/tw/waterball/judgegirl/problem/domain/usecases/DeleteTestcaseUseCase.java [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+@Named
+public class DeleteTestcaseUseCase {
+
+    private ProblemRepository testCaseRepository;
+
+    public DeleteTestcaseUseCase(ProblemRepository testCaseRepository) {
+        this.testCaseRepository = testCaseRepository;
+    }
+
+    public void execute(Request request) throws NotFoundException {
```
Comment (Johnny850807) [inline Domain/Problem/src/main/java/tw/waterball/judgegirl/problem/domain/usecases/DeleteTestcaseUseCase.java:23]: no NotFoundException is thrown

File: Domain/Problem/src/main/java/tw/waterball/judgegirl/problem/domain/usecases/DeleteTestcaseUseCase.java [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+/**
+ * @author swshawnwu@gmail.com(ShawnWu)
+ */
+
+@Named
+public class DeleteTestcaseUseCase {
+
+    private ProblemRepository testCaseRepository;
+
+    public DeleteTestcaseUseCase(ProblemRepository testCaseRepository) {
```
Comment (Johnny850807) [inline Domain/Problem/src/main/java/tw/waterball/judgegirl/problem/domain/usecases/DeleteTestcaseUseCase.java:19]: use lombok

**Review by Johnny850807 — CHANGES_REQUESTED** [review]
Review done

**Review by Wally5077 — COMMENTED** [review]
Review done

**Review by Wally5077 — APPROVED** [review]
LGTM

**Review by Johnny850807 — APPROVED** [review]
LGTM

**Pull request merged**